Userland/DevTools/UserspaceEmulator/SoftFPU.h

```diff
@@ -0,0 +1,603 @@
+/*
+ * Copyright (c) 2020, Andreas Kling <kling@serenityos.org>
+ * Copyright (c) 2020, Leon Albrecht <loen2002.la@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include "Report.h"
+#include <LibX86/Instruction.h>
+#include <LibX86/Interpreter.h>
+
+#include <math.h>
+#include <string.h>
+
+namespace UserspaceEmulator {
+
+class Emulator;
+class SoftCPU;
+
+class SoftFPU final {
+public:
+    SoftFPU(Emulator& emulator, SoftCPU& cpu)
+        : m_emulator(emulator)
+        , m_cpu(cpu)
+    {
+    }
+
+    bool c0() const { return m_fpu_c0; }
+    bool c1() const { return m_fpu_c1; }
+    bool c2() const { return m_fpu_c2; }
+    bool c3() const { return m_fpu_c3; }
+
+    void set_c0(bool val) { m_fpu_c0 = val; }
+    void set_c1(bool val) { m_fpu_c1 = val; }
+    void set_c2(bool val) { m_fpu_c2 = val; }
+    void set_c3(bool val) { m_fpu_c3 = val; }
+
+    long double fpu_get(u8 index) const;
+
+    void fpu_push(long double value);
+    long double fpu_pop();
+    void fpu_set_absolute(u8 index, long double value);
+    void fpu_set(u8 index, long double value);
+
+    template<typename T>
+    T* mmx_get(u8 index);
+    template<typename T>
+    const T* mmx_get(u8 index) const;
+    template<typename T>
+    void mmx_set(u8 index, T* values);
+    void mmx_set(u8 index, u64 value);
+
+private:
+    friend class SoftCPU;
+
+    Emulator& m_emulator;
+    SoftCPU& m_cpu;
+
+    enum class FPU_Exception : u8 {
+        InvalidOperation = 0b000'0001,
+        DenormalizedOperand = 0b0000'0010,
+        ZeroDivide = 0b0000'0100,
+        Overflow = 0b0000'1000,
+        Underflow = 0b0001'0000,
+        Precision = 0b0010'0000,
+        StackFault = 0b0100'0001, // Implies InvalidOperation
+    };
+
+    enum class FPU_Tag : u8 {
+        Valid = 0x00,
+        Zero = 0x01,
+        Special = 0x10,
+        Empty = 0x11
+    };
+
+    enum class RoundingMode : u8 {
+        NEAREST = 0x00,
+        DOWN = 0x01,
+        UP = 0x10,
+        TRUNK = 0x11
+    };
+
+    // FIXME: can we make this more a constexpr?
```

Comment:
This function doesn't look like a good `constexpr` candidate. What am I missing?

Reply:
its only called with constexpr Arguments, so its technically possible to optimize the if statements away, would ALWAYS_INLINE solve that?